src/protocols/two_party_schnorr/mu_sig/mod.rs

```diff
@@ -1,20 +1,264 @@
+
+        let mut apk_vec_2_n = apk_vec.clone();
+        let pk1 = apk_vec_2_n.remove(0);
+        let sum = apk_vec_2_n
+            .iter()
+            .fold(pk1, |acc, pk| acc.combine(&ec_context, pk).unwrap());
+
+        KeyAgg {
+            apk: sum,
+            hash: hash_vec[*party_index].clone(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EphemeralKey {
+    pub keypair: KeyPair,
+    pub commitment: BigInt,
+    pub blind_factor: BigInt,
+}
+
+impl EphemeralKey {
+    pub fn create(ec_context: &EC) -> EphemeralKey {
+        let keypair = KeyPair::create(ec_context);
+        let (commitment, blind_factor) =
+            HashCommitment::create_commitment(&keypair.public_key.to_point().x);
+        EphemeralKey {
+            keypair,
+            commitment,
+            blind_factor,
+        }
+    }
+
+    pub fn test_com(r_to_test: &PK, blind_factor: &BigInt, comm: &BigInt) -> bool {
+        let computed_comm = &HashCommitment::create_commitment_with_user_defined_randomness(
+            &r_to_test.to_point().x,
+            blind_factor,
+        );
+        computed_comm == comm
+    }
+
+    pub fn add_ephemeral_pub_keys(ec_context: &EC, r1: &PK, r2: &PK) -> PK {
+        r1.combine(ec_context, r2).unwrap()
+    }
+
+    pub fn hash_0(r_hat: &PK, apk: &PK, message: &[u8]) -> BigInt {
+        HSha256::create_hash(vec![
+            &BigInt::from(0),
+            &r_hat.to_point().x,
+            &apk.to_point().x,
+            &BigInt::from(message),
+        ])
+    }
+
+    pub fn add_signature_parts(s1: &BigInt, s2: &BigInt, r_tag: &PK) -> (PK, BigInt) {
+        (*r_tag, BigInt::mod_add(&s1, &s2, &EC::get_q()))
+    }
+
+    pub fn verify(
+        ec_context: &EC,
+        signature: &BigInt,
+        r_tag: &PK,
+        apk: &PK,
+        message: &[u8],
+    ) -> Result<(), ProofError> {
+        let c = HSha256::create_hash(vec![
+            &BigInt::from(0),
+            &r_tag.to_point().x,
+            &apk.to_point().x,
+            &BigInt::from(message),
+        ]);
+        let mut s_g = PK::to_key(ec_context, &EC::get_base_point());
+
+        let mut c_y = *apk;
+        assert!(
+            c_y.mul_assign(ec_context, &SK::from_big_int(ec_context, &c))
+                .is_ok()
+        );
+        assert!(
+            s_g.mul_assign(ec_context, &SK::from_big_int(ec_context, signature))
+                .is_ok()
+        );
+
+        if s_g == r_tag.combine(ec_context, &c_y).unwrap() {
+            Ok(())
+        } else {
+            Err(ProofError)
+        }
+    }
+
+    pub fn sign(r: &EphemeralKey, c: &BigInt, x: &KeyPair, a: &BigInt) -> BigInt {
+        BigInt::mod_add(
+            &r.keypair.private_key.to_big_int(),
+            &BigInt::mod_mul(
+                c,
+                &BigInt::mod_mul(&x.private_key.to_big_int(), a, &EC::get_q()),
+                &EC::get_q(),
+            ),
+            &EC::get_q(),
+        )
+    }
+}
+
+pub mod party_two {
```

Comment:
party 2 is redundent. The only function in this mod is `key_aggregation` which can be replaced by function `key_aggregation_n `. See line 138 in test.rs